sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala

```diff
@@ -309,58 +360,46 @@ class MempoolMap(
-    // First all the txs with nonce too low are discarded. The subsequent txs don't need to be checked for nonce
-    // because they are ordered by increasing nonce. They are candidate to be added by default to the execTxs map, because
-    // they come from reverted blocks so it is impossible to have nonce gaps. They still need to be checked for balance.
-    // If a tx is found invalid for balance, it is discarded. All the subsequent txs, if valid, will become not executable
-    // and they will be added to the nonExec map.
-    var destMap = newExecTxs
-    var haveBecomeNonExecutable = false
 
     if (existRejectedTxsWithValidNonce(txsFromRejectedBlocks, newExpectedNonce)) {
-      txsFromRejectedBlocks.withFilter(tx => tx.getNonce.compareTo(newExpectedNonce) >= 0).foreach { tx =>
-        if (tx.size() <= MaxTxSize && balance.compareTo(tx.maxCost) >= 0) {
-          all.put(tx.id, tx)
-          destMap.put(tx.getNonce, tx.id)
-        } else {
-          if (!haveBecomeNonExecutable) {
-            destMap = newNonExecTxs
-            haveBecomeNonExecutable = true
-          }
-        }
-      }
-
+      restoreRejectedTransactions(account, txsFromRejectedBlocks.dropWhile(tx => tx.getNonce.compareTo(newExpectedNonce) < 0))
     }
-
-    // First all the txs with nonce too low are discarded. The subsequent txs don't need to be checked for nonce
-    // because they are ordered by increasing nonce. They are candidate to be added by default to the execTxs map, unless
-    // a previous tx was found invalid. They remaining txs are checked for balance.
-    // If a tx is found invalid, it is removed from the mem pool. All the subsequent txs, if valid, will become not executable
-    // and they will be added to the nonExec map.
-    val execTxsOpt = executableTxs.remove(account)
-    if (execTxsOpt.nonEmpty) {
-      execTxsOpt.get
-        .withFilter {
-          case (nonce, id) => {
-            if (nonce.compareTo(newExpectedNonce) < 0) {
-              all.remove(id)
-              false
-            } else
-              true
-          }
-        }.foreach { case (nonce, id) =>
+    else {
+      //The new expected nonce is at least the same as the old one, so I don't need to check the remaining txs
+      //for the nonce gap. No need to check for tx size too, because this check was already performed when they were added
+      // the first time in the mempool
+      val fromAddress = account.asInstanceOf[AddressProposition].address()
+      val balance = accountStateReaderProvider.getAccountStateReader().getBalance(fromAddress)
+
+      val newExecTxs: mutable.TreeMap[BigInteger, ModifierId] = new mutable.TreeMap[BigInteger, ModifierId]()
+      val newNonExecTxs: mutable.TreeMap[BigInteger, ModifierId] = new mutable.TreeMap[BigInteger, ModifierId]()
+
+      // Recreates from scratch the account's nonExecTxs and execTxs maps, starting from txs from rejected blocks.
```

Comment:
this comment seems outdated.
Here we don't have any rejected txs, but only re-validate the existing mempool txs

Reply:
Done